
---
title: "Assignment 2"
---

```{r}
library(tidyverse)
```

### Clinical cancer data

**6. How many samples in the METABRIC dataset have high cellularity but have no recorded classification with the 3-gene classifier?**

```{r}
metabric <- read_tsv("brca_metabric_clinical_data.tsv")
metabric

table(metabric$Cellularity)
table(metabric$`3-Gene classifier subtype`)

sum(metabric$Cellularity == "High")
sum(metabric$Cellularity == "High" & !is.na(metabric$Cellularity))
sum(metabric$Cellularity == "High", na.rm = TRUE)

sum(is.na(metabric$`3-Gene classifier subtype`))

sum(metabric$Cellularity == "High" & !is.na(metabric$Cellularity) & is.na(metabric$`3-Gene classifier subtype`))

sum(metabric$Cellularity == "High" & is.na(metabric$`3-Gene classifier subtype`), na.rm = TRUE)

# note dplyr filter that we'll introduce in a couple of weeks is much easier and
# handles the missing value more like how we'd expect
```

*Hint:* this is actually quite tricky -- you may want to check for missing
values in the cellularity column as well as the column for the 3-gene
classifier.

---

### Working with data from an Excel spreadsheet

**7. Read into R one of your own Excel spreadsheets and explore the data using `summary()`, `table()` and/or `plot()`**

We haven't yet showed how to read Excel spreadsheets into R. There are more than
package providing this capability but for this exercise use the `readxl` package
from the tidyverse.

`readxl` should have been installed as part of the tidyverse but if you find
that's not the case, you will have to install it separately.

`readxl` is not loaded as one of the core tidyverse packages when you run
`library(tidyverse)` so you'll have to load this separately.

If you don't have an Excel spreadsheet to hand, you can use the clinical
annotation data for the NCI-60 cell line data set (https://bioinformatics-core-shared-training.github.io/r-intro/data/cellline_nci60_clinical_data.xlsx).

```{r}
library(readxl)

nci60 <- read_excel("cellline_nci60_clinical_data.xlsx", na = "NA")
# need to comment on the na argument when reading data in (blanks assigned to NA
# by default, not "NA" strings)

nci60

# summary of each variable
summary(nci60)

# note that categorical variables are not factors so would need converting
# using as.factor() or for this to be specified when reading the data in
# nevertheless we can run table() on these
table(nci60$`Cancer Type`)

# scatter plot to see if there is a relationship between the number of point
# mutations and the fraction of the genome altered
plot(nci60$`Mutation Count`, nci60$`Fraction Genome Altered`)
```

---

### Create a report for your assignment

Click on the 'Knit' menu at the top of this file and select either whichever
option you prefer to create an HTML, PDF or Word document version of your
assignment. This will run all the code chunks and "knit" the resulting
results with the surrounding text to produce a report.

#### Additional markdown features

**_Emphasis_**

Can use either underscores or asterisks and combination of both for both bold
and italic emphasis.

```
*italic*   **bold**
_italic_   __bold__
```

**_Inline Code_**

Use of backticks (`).

```
`mean(starwar$height)`
```

**_Horizontal Rule_**

Three of more astisks or dashes.

```
*******

-------
```

**_Plain Code Blocks_**

Plain code blocks are displayed in a fixed-width font but not evaluated.

```
This text is displayed verbatim / preformatted
```

Have a look at examples of these in the text parts of this R markdown document
and how these are rendered when you run the knit operation.

For more features, see the help page in RStudio that can be accessed from the
the help menu: **Help > Markdown Quick Reference**

---
